### PR TITLE
utils/lister: Limit the API of scan_dir() to fs::path

### PR DESCRIPTION
**Pull request by dawmd**
Callers of lister::scan_dir() should perform the conversion to std::filesystem::path explicitly.

Continuation of #15165

### CODE REVIEW

File: db/hints/manager.cc [hunk 1]
```diff
@@ -1288,7 +1288,7 @@ void manager::rebalance_segments_for(
     }
 
     for (unsigned i = 0; i < smp::count && !segments_to_move.empty(); ++i) {
-        fs::path shard_path_dir(fs::path(hints_directory.c_str()) / seastar::format("{:d}", i).c_str() / ep.c_str());
+        fs::path shard_path_dir(hints_directory / seastar::format("{:d}", i).c_str() / ep.c_str());
```
Comment (tchaikov) [inline db/hints/manager.cc:1291]: note to myself, we don't need to use `seastar::format()` here. `fmt::to_string()` would be better.

File: db/hints/manager.cc [hunk 2]
```diff
@@ -1391,7 +1391,7 @@ class directory_initializer::impl {
 
         return with_semaphore(_lock, 1, [this] () {
             manager_logger.debug("Rebalancing hints in {}", _hints_directory);
-            return manager::rebalance(_hints_directory).then([this] {
+            return manager::rebalance(fs::path{_hints_directory}).then([this] {
```
Comment (tchaikov) [inline db/hints/manager.cc:1394]: could go further by adding a commit to change `_hints_directory` to `fs::path`.

**Review by tchaikov — COMMENTED** [review]
couple nits on the commit message and its title:

> Limit the API of utils/lister.hh to fs::path 

if you take at the output of `git log db/hints/manager.cc`, quite a few titles of of the commit messages have prefixes for the areas touched by the certain commit. the prefix helps developer to have a quick indication which area(s) of a commit is changing. in this case, i'd suggest at least include "hints" and "lister" in the title.

> Callers of lister::scan_dir() should perform the conversion to std::filesystem::path explicitly.

the commit message summarizes the change. it reads like an action item. but what problem does it address? does it fix a bug? or does it improve the readability? or prepares for yet another change? in other words, it fails to explain why.

**Review by denesb — APPROVED** [review]
Note that we usually prepend the affected subsystem or even file to the commit message, e.g.:

    utils/lister: Limit the API of utils/lister.hh to fs::path

**Comment by dawmd** [comment]
Thank you both for the suggestions about commit messages. I'll edit it in a moment, also changing the `directory_initializer::_hint_directory`'s type.

**Comment by dawmd** [comment]
Let me know if the commit message could be improved further. The changes and their purpose seem pretty straightforward to me, but chances are I'm still missing the whole point. I'll be happy to learn.

**Pull request closed without merging**